sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala

```diff
@@ -43,6 +43,22 @@ object EliminateResolvedHint extends Rule[LogicalPlan] {
       case h: ResolvedHint =>
         hintErrorHandler.joinNotFoundForJoinHint(h.hints)
         h.child
+      case join: Join if join.condition.isDefined =>
+        join.copy(
+          hint = join.hint.copy(
+            leftHint = removeCartesianProductJoinHint(join.hint.leftHint),
+            rightHint = removeCartesianProductJoinHint(join.hint.rightHint)))
+    }
+  }
+
+  def removeCartesianProductJoinHint(hint: Option[HintInfo]): Option[HintInfo] = {
```

Comment:
I think we should pass a correct condition (`leftKeys` and `rightKeys`) into `CartesianProductExec` instead of removing the hint: 
https://github.com/apache/spark/blob/ac6406e7571027ba61193452d7a54f7895d2cbdc/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala#L202

Reply:
It seems the spark strategy incrrectly removes `a#0 = a#2`; 
```
== Optimized Logical Plan ==
Sort [a#0 ASC NULLS FIRST], true
+- Join Inner, (a#0 = a#2), leftHint=(strategy=shuffle_replicate_nl)
   :- Filter isnotnull(a#0)
   :  +- Relation[a#0,b#1] parquet
   +- Filter isnotnull(a#2)
      +- Relation[a#2,b#3] parquet

== Physical Plan ==
*(3) Sort [a#0 ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(a#0 ASC NULLS FIRST, 200), true, [id=#87]
   +- CartesianProduct
      :- *(1) Project [a#0, b#1]
      :  +- *(1) Filter isnotnull(a#0)
      :     +- *(1) ColumnarToRow
      :        +- FileScan parquet default.test4[a#0,b#1] Batched: true, DataFilters: ...
      +- *(2) Project [a#2, b#3]
         +- *(2) Filter isnotnull(a#2)
            +- *(2) ColumnarToRow
               +- FileScan parquet default.test5[a#2,b#3] Batched: true, DataFilters: ...
```

Reply:
> I think we should pass a correct condition (`leftKeys` and `rightKeys`) into `CartesianProductExec` instead of removing the hint:
> https://github.com/apache/spark/blob/ac6406e7571027ba61193452d7a54f7895d2cbdc/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala#L202

Yea, in default Cartesian Product Join situation,  it didn't need condition at all.  So in default, it seems don't have condition when build data.

Reply:
@maropu See latest change, it's ok to do like this?

Reply:
Since we don't know the join keys's condition is `EqualTo` or `EaultNullSafe` so it's better just not remove it in `ExtractEqualJoinKeys`

Reply:
You cannot use the original condition in logical.Join?
```
case p @ ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right, hint) =>
  p.condition <-- This?
```

Reply:
> You cannot use the original condition in logical.Join?
> 
> ```
> case p @ ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right, hint) =>
>   p.condition <-- This?
> ```

Don't know we can write like this....updated..